### ISSUE TITLE
hyperlinks that span lines make all of both lines clickable

### ISSUE THREAD

**Issue by mslgam** [wkhtmltopdf/wkhtmltopdf#2422]
When a hyperlink (`<a href...>`) spans a line, clicking anywhere in either of the lines will activate the link.

I'm not using any command-line options (just 'wkhtmltopdf in.html out.pdf').

I'm running it under Kubuntu 14.04, and here's some more detail:
wkhtmltopdf -V outputs:
  wkhtmltopdf 0.12.1 (with patched qt)
uname -srvp outputs:
  Linux 3.13.0-24-generic #46-Ubuntu SMP Thu Apr 10 19:11:08 UTC 2014 x86_64

Here's an HTML file that demonstrates this bug:
```html
<!DOCTYPE html>
<html lang="en">
  <body>
    <p>This is a long paragraph that will wrap. This text should be on the first line.
      <a href="http://www.google.com/">Here's a link that crosses a line.</a>
      <a href="http://www.yahoo.com/">Here's a link on the second line.</a>
      Here's some unlinked text.
    </p>
  </body>
</html>
```

A workaround I've found is to replace all spaces in the link text with ```&nbsp;```'s so the link doesn't wrap.

**Comment by ashkulz** [comment]
Thanks for the very good bug report, I can confirm this is the case even with the development version.

**Comment by ashkulz** [comment]
Looks like `QWebPrinter::elementLocation()` uses `RenderObject.absoluteBoundingBoxRect()` which computes a bounding box that is the union of all the rectangles enclosing the text. In this pathological case, it means the complete area of the two lines. To really fix it, the signature of `QWebPrinter::elementLocation()` will have to be changed which will have to be thought about in detail.

**Comment by mslgam** [comment]
Maybe the link could be split into two pieces (two separate hyperlinks pointing to the same destination), one per line?

**Comment by ashkulz** [comment]
Yeah, that's what I meant by changing the signature -- it would have to report multiple non-overlapping areas separately instead of a single area i.e. two segments of each line instead of a single block. If I'm going to change that, might as well look at other issues in this area (even some that you reported :smile:) before changing the approach.

**Comment by ktpsu** [comment]
Having problems with this as well. I'm also seeing that two links on the same line makings the whole line clickable... but to only one of the links.

**Comment by cmr5289** [comment]
Hey, I would like to take a look at this or help fix the issue as I currently use this project. I have looked into the QWebPrinter::elementLocation() and was unable to really pinpoint where that is in the files and how to go about addressing this issue.

**Comment by claweyenuk** [comment]
This issue happens in Windows 10 as well.

**Comment by nielsbox** [comment]
Bump, any update on this? @ashkulz

**Comment by nategasser** [comment]
Not a fix, but an ok workaround -- I used PHP DOM to preprocess the HTML before sending to wkhtmlpdf.  If you split the text of each link on space, then link each word separately, concat them, and replace the original link, it works.  Put a space after each word (except the last) and (at least in my tests) you can't even tell by looking at the PDF.

But yeah, a fix would be good :)
